// @tanstack/start@1.76.1/dist/esm/client/Scripts.js
import { jsx, Fragment } from "react/jsx-runtime";
import { createElement } from "react";
import { useRouter, useRouterState } from "@tanstack/react-router";
import { Asset } from "./Asset.js";
const Scripts = () => {
  const router = useRouter();
  const assetScripts = useRouterState({
    select: (state) => {
      const assetScripts2 = [];
      state.matches.map((match) => router.looseRoutesById[match.routeId]).forEach(
        (route) => {
          var _a, _b, _c;
          return (_c = (_b = (_a = router.manifest) == null ? void 0 : _a.routes[route.id]) == null ? void 0 : _b.assets) == null ? void 0 : _c.filter((d) => d.tag === "script").forEach((asset) => {
            assetScripts2.push({
              tag: "script",
              attrs: asset.attrs,
              children: asset.children
            });
          });
        }
      );
      return assetScripts2;
    }
  });
  const { scripts } = useRouterState({
    select: (state) => ({
      scripts: state.matches.map((match) => match.scripts).filter(Boolean).flat(1).map(({ children, ...script }) => ({
        tag: "script",
        attrs: {
          ...script,
          suppressHydrationWarning: true
        },
        children
      }))
    })
  });
  const allScripts = [...scripts, ...assetScripts];
  return /* @__PURE__ */ jsx(Fragment, { children: allScripts.map((asset, i) => (
    // eslint-disable-next-line @eslint-react/no-array-index-key
    /* @__PURE__ */ createElement(Asset, { ...asset, key: `tsr-scripts-${asset.tag}-${i}` })
  )) });
};
export {
  Scripts
};
